File: DailyChallenges/Scala/LeetCode_2300.scala
object Solution {
    def successfulPairs(spells: Array[Int], potions: Array[Int], success: Long): Array[Int] = {
        val p = potions.sorted

        def bs(t: Int, l:Int = 0, r: Int = p.length): Int ={
            if (l>=r) return l
            val m = l + (r-l)/2;
            if (p(m)*t.toLong >= success) bs(t, l, m) else bs(t, m+1, r)
        }
        spells.map(sp => p.length - bs(sp))
    }
}

File: DailyChallenges/CPP/LeetCode_2300.cpp
class Solution {
public:
    vector<int> successfulPairs(vector<int>& spells, vector<int>& potions, long long success) {
        sort(potions.begin(), potions.end());
        vector<int> ans;
        for (int sp: spells){
            auto itr = lower_bound(potions.begin(), potions.end(), (success+sp-1)/sp);
            ans.push_back(potions.end()-itr);
        }
        return ans;
    }
};
